Put memory tips in central place in annex (#65)
moj-analytical-services/user-guidance: source/documentation/tools.md

To find out if you have hit the memory limit, you can check [Grafana](https://grafana.services.alpha.mojanalytics.xyz/login). For more about that, and description of how memory is allocated on the Analytical Platform, see: [Memory limits](annexes.html#memory-limits).

To understand what is taking up the memory, you can try [pryr](https://rdrr.io/cran/pryr/man/mem_used.html). To free up memory, for example when a variable points to a big dataframe, you can instead assign something a null to the variable, and then run [`gc()`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/gc).

To free up memory, it's often easiest to [clear your RStudio session](#clearing-your-rstudio-session).

### Clearing your RStudio session

The RStudio session is the short term 'state' of RStudio, including:

* which project and files are open (displayed in the RStudio interface)
* console and terminal history
* the global environment – R variables that have been set and are held in memory

The session is persisted between restarts of RStudio, to make it easier for you to pick up where you left off. However you can clear it to solve a number of problems, including if the memory is full.

> To clear the RStudio session:

Author: Nice tip - I've done this in a separate PR: https://github.com/moj-analytical-services/user-guidance/pull/69

* "Restart" - Often problems with the tool can be solved by restarting the software on the server.
* "Upgrade" - Another release of the tool is available. Occasionally new versions of R Studio are made available on the Analytical Platform. In this case all users will be given the opportunity to upgrade on the control panel. New versions provide new features and bugfixes to the tool. In addition, there some releases come with improvements to the way RStudio is containerized and integrated with Analytical Platform. You should aim to upgrade when it is offered, although just in case it causes minor incompatibilities with your R code, you should not do it in the days just before you have a critical release of your work. When pressed, the status will change to 'Deploying' and then 'Upgraded'. The __Upgrade__ button will no longer be visible (until another version becomes available).

## Using RStudio

> For general guidance in using RStudio, see the [RStudio documentation](https://docs.rstudio.com/)

Reviewer: ```suggestion
For general guidance in using RStudio, see the [RStudio documentation](https://docs.rstudio.com/).
```

### RStudio memory issues

> It is common to run out of memory in RStudio.

Reviewer: I wonder if users understand that everything is stored in memory in R, which is what leads to high utilisation. Is there more we can do to encourage good memory management from the outset? I feel like people may think it's an issue with the AP, whereas it's actually an issue with their code.

> To find out if you have hit the memory limit, you can check [Grafana](https://grafana.services.alpha.mojanalytics.xyz/login). For more about that, and description of how memory is allocated on the Analytical Platform, see: [Memory limits](annexes.html#memory-limits).

Reviewer: ```suggestion
To find out if you have hit the memory limit, you can check [Grafana](https://grafana.services.alpha.mojanalytics.xyz/login). For more about that, and description of how memory is allocated on the Analytical Platform, see the [memory limits](annexes.html#memory-limits) section.
```

> To understand what is taking up the memory, you can try [pryr](https://rdrr.io/cran/pryr/man/mem_used.html). To free up memory, for example when a variable points to a big dataframe, you can instead assign something a null to the variable, and then run [`gc()`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/gc).

Reviewer: ```suggestion
To understand what is taking up the memory, you can us the [pryr](https://rdrr.io/cran/pryr/man/mem_used.html) package. To free up memory, for example when a variable points to a big dataframe, you can instead assign something null to the variable and then run [`gc()`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/gc).
```

To free up memory, it's often easiest to [clear your RStudio session](#clearing-your-rstudio-session).

### Clearing your RStudio session

The RStudio session is the short term 'state' of RStudio, including:

* which project and files are open (displayed in the RStudio interface)

> * Console and Terminal's history
> * the 'Global Environment' - R variables that have been set and are held in memory

Reviewer: ```suggestion
* console and terminal history
* the global environment – R variables that have been set and are held in memory
```

The session is persisted between restarts of RStudio, to make it easier for you to pick up where you left off. However you can clear it to solve a number of problems, including if the memory is full.

To clear the RStudio session:

> 1. Close RStudio, if it is open in any window (because continually saves its session to disk).

Reviewer: ```suggestion
1. Close RStudio, if it is open in any window (because it continually saves its session to disk).
```

> 2. Open JupyterLab. In JupyterLab open a Terminal and run:
> 
>    ```bash
>    rm -rf ~/.rstudio.bak; mv ~/.rstudio/ ~/.rstudio.bak
>    ```
> 

Reviewer: ```suggestion
2. Open JupyterLab.
3. In JupyterLab, open a Terminal and run:

   ```bash
   rm -rf ~/.rstudio.bak; mv ~/.rstudio/ ~/.rstudio.bak
   ```

```

> 3. In Control Panel click "Restart" for RStudio:

Reviewer: ```suggestion
3. In the control panel, select the __Restart__ button for RStudio.
```

   ![RStudio's "Restart" button in Control Panel](images/tools/restart_rstudio.png)

> 4. In Control Panel click "Open" for RStudio. It takes 1-5 minutes to be running again, and then refreshing the browser should show RStudio.

Reviewer: ```suggestion
4. In the control panel, select __Open__ for RStudio. It maybe take between one and five minutes before RStudio is available. You may need to refresh your browser for the tool to load.
```

but because we are using the Open Source version of R Studio, there is only a
single conda environment available. This means having to be careful to make sure packages don't pollute your environment from another project. The following commands can be used to manage your environments.

#### Reset your conda environment to default

> This will delete packages that the user has installed your "rstudio" conda environment, leaving only the base packages:

Reviewer: ```suggestion
This will delete packages that you have installed in your `rstudio` conda environment, leaving only the base packages:
```

```bash
conda env export -n base| grep -v "^prefix: " > /tmp/base.yml && conda env update --prune -n rstudio -f /tmp/base.yml && rm /tmp/base.yml
```

It is recommended to do this before starting a new project, to ensure that no unused dependencies are exported when you export an `environment.yml` for this project.

#### Hard reset of your conda environment

> This will completely delete your "rstudio" conda environment, and recreate it with the base packages:

Reviewer: ```suggestion
This will completely delete your `rstudio` conda environment, and recreate it with the base packages:
```

2. In Control Panel, for R Studio, select the "Restart" button

> It can be useful to do this if you have tried [Reset your conda environment to default](#reset-your-conda-environment-to-default) and are still having problems.

Reviewer: ```suggestion
It can be useful to do this if you have tried to [reset your conda environment to default](#reset-your-conda-environment-to-default) and are still having problems.
```
